### DISCUSSION TITLE
No next track when card removed

### DISCUSSION THREAD

**Discussion by oxygi** [MiczFlor/RPi-Jukebox-RFID#2243] [Q&A] [answered]
I have successfully built a working box for my disabled daughter. But unfortunately there are two things that should be implemented for mentally handicapped people.
I have installed an RC522 reader with a RotaryEncoder for very simple control. Clicking on the RotaryEncoder starts the next song.
1) The music should only play when a card is present. I have already achieved this with the settings. The music stops when the card has been removed. However, if there are several tracks on the card, the card is removed from the reader and you then press the RotaryEncoder, the box plays the music even without a card. Can this be changed?
2) My daughter needs more instruction when using the box. The switch-on sound and switch-off sound are already very good, but in my opinion two sounds are missing here: one when the card has been scanned successfully and one when there is no more track on the card when she presses the RotaryEncoder. Would this be technically feasible?

[RotaryEncoder_Press]
type : Button
pin : 13
pull_up_down : pull_up
functioncall : functionCallPlayerNext
hold_mode: SecondFunc
hold_time: 3.0
functionCall2: functionCallShutdown

[RotaryEncoder_Rotate]
type : RotaryEncoder
timebase : 0.02
pin1 : 5
pin2 : 6
functioncall1 : functionCallVolU
functioncall2 : functionCallVolD

I am on Version [v2.5.0]
Thank you very much in advance!

**Comment by AlvinSchiller** [comment]
Hi, and welcome :)
1. No this behavior cannot be changed right now. The button functions (like the WebUi) do not check the state of cards even with "placing" activated. This could possibly be added, butwould take some effort and use case definitions.
2.  The extras sound should be not that difficult to add, though the sounds itself are needed / have to be created. 

Note that we do not plan to add more feature to the version 2, in favor for version 3.
Would it be possible for you to update to v3? Or are there other features missing that make v2 mandatory?

I would suggest you create a feature request (future3) for both additions, so they can be discussed and picked up :)
Do you have any experience in coding so we can guide you adding this yourself?

**Reply by oxygi** [reply]
Thanks for the help - but even if I change the version, the "feature" would still come later... I understand a little bit the structure of the folders and also see if-statements :-) Unfortunately that's all. I have already found the sound for scanning the card on the Internet. I would record the sound for the end of the card myself so that my daughter knows what to do. Unfortunately, the box has to explain a little more here compared to other people ;-) From my feeling I would need some place in the code where I insert the sound similar to this line of code: `/usr/bin/mpg123 /home/pi/RPi-Jukebox-RFID/shared/shutdownsound.mp3`

**Reply by oxygi** [reply]
Quick question, as the situation has already come up again this morning: Is there no way to deactivate the NextTrack or the button if there is no card on it? She always presses the button and it starts even without a music card. :-(

**Reply by AlvinSchiller** [reply]
> Quick question, as the situation has already come up again this morning: Is there no way to deactivate the NextTrack or the button if there is no card on it? She always presses the button and it starts even without a music card. :-(

Please be aware that this is a hack and will break some functionality like playback from the WebUI!

Edit `/home/pi/RPi-Jukebox-RFID/scripts/playout_controls.sh` and change line 671 to this:
```
        PLAYSTATE=$(echo -e "status\nclose" | nc -w 1 localhost 6600 | grep -o -P '(?<=state: ).*')
        # Only execute when currently playing
        if [ "$PLAYSTATE" == "play" ]; then
            mpc next
        fi
```

**Reply by AlvinSchiller** [reply, marked as answer]
> Thanks for the help - but even if I change the version, the "feature" would still come later... I understand a little bit the structure of the folders and also see if-statements :-) Unfortunately that's all. I have already found the sound for scanning the card on the Internet. I would record the sound for the end of the card myself so that my daughter knows what to do. Unfortunately, the box has to explain a little more here compared to other people ;-) From my feeling I would need some place in the code where I insert the sound similar to this line of code: `/usr/bin/mpg123 /home/pi/RPi-Jukebox-RFID/shared/shutdownsound.mp3`

### Sound on RFID Scan
For this you could edit the file `/home/pi/RPi-Jukebox-RFID/scripts/daemon_rfid_reader.py` and add the following before the `subprocess` call for `rfid_trigger_play`. Provide the file `shared/rfidscansound.mp3` with your desired sound.
```
subprocess.call(['/usr/bin/mpg123 ' + dir_path + '/../shared/rfidscansound.mp3'], shell=True)
```

### Sound on playlist end
~~This would get more complicated to add to the code, and i have no idea for this right now.
But if you record an instruction anyway, would it be possible to just add it as track to the end of the playlist? It wouldn't be played if she presses `next` again though :/~~

Add this after the addition "deactivate the NextTrack" in `/home/pi/RPi-Jukebox-RFID/scripts/playout_controls.sh` for `playernext`.
This reads the playstate again after the "next" command. Provide the file shared/endofplaylist.mp3 with your desired sound.

```
        PLAYSTATE=$(echo -e "status\nclose" | nc -w 1 localhost 6600 | grep -o -P '(?<=state: ).*')
        # Only execute when currently stopped
        if [ "$PLAYSTATE" == "stop" ]; then
            /usr/bin/mpg123 /home/pi/RPi-Jukebox-RFID/shared/endofplaylist.mp3
        fi
```


So both would look like this:

```
        PLAYSTATE=$(echo -e "status\nclose" | nc -w 1 localhost 6600 | grep -o -P '(?<=state: ).*')
        # Only execute when currently playing
        if [ "$PLAYSTATE" == "play" ]; then
            mpc next
        fi
        PLAYSTATE=$(echo -e "status\nclose" | nc -w 1 localhost 6600 | grep -o -P '(?<=state: ).*')
        # Only execute when currently stopped
        if [ "$PLAYSTATE" == "stop" ]; then
            /usr/bin/mpg123 /home/pi/RPi-Jukebox-RFID/shared/endofplaylist.mp3
        fi
```

**Reply by oxygi** [reply]
It all works perfectly - for me, it doesn't affect the web player. I think it's really easy to understand even with the web player. This should find its way into the official programming as a parameter! Thank you very very much for your effort!

**Comment by pabera** [comment]
If you use future3, you can achieve this as well. If you connect a buzzer to your Phoniebox, you can make it Buzz depending on the state. [Here is the code that explains it in depth](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/develop/src/jukebox/components/gpio/gpioz/plugin/connectivity.py). There is more functionality as well. You can simply configure it in the `gpio.yaml` file.

https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/6e9e4f7ae1db3a83ba9e147e320f1aaa6499de70/resources/default-settings/gpio.example.yaml#L6-L23
